File: frontend/src/types/recommendations.ts
/**
 * Recommendation Types - AI-powered next-best-actions
 */

// Enums
export type RecommendationType =
  | "contact_action"
  | "deal_action"
  | "follow_up"
  | "outreach"
  | "engagement"
  | "upsell"
  | "retention"
  | "prioritization";

export type RecommendationPriority = "critical" | "high" | "medium" | "low";

export type RecommendationStatus =
  | "pending"
  | "accepted"
  | "dismissed"
  | "completed"
  | "expired";

// Main types
export interface Recommendation {
  id: string;
  contact_id: string | null;
  deal_id: string | null;
  type: RecommendationType;
  priority: RecommendationPriority;
  status: RecommendationStatus;
  title: string;
  description: string;
  reasoning: string;
  suggested_action: string;
  action_template: string | null;
  confidence: number;
  impact_score: number;
  urgency_score: number;
  context_data: Record<string, unknown>;
  created_at: string;
  expires_at: string | null;
}

export interface RecommendationList {
  recommendations: Recommendation[];
  total: number;
  pending_count: number;
  high_priority_count: number;
}

export interface NextBestActions {
  contact_actions: Recommendation[];
  deal_actions: Recommendation[];
  follow_ups: Recommendation[];
  top_priorities: Recommendation[];
  generated_at: string;
}

export interface RecommendationActionRequest {
  action: "accept" | "dismiss" | "complete";
  notes?: string;
}

// Helper functions
export function getPriorityColor(priority: RecommendationPriority): string {
  switch (priority) {
    case "critical":
      return "text-red-600";
    case "high":
      return "text-orange-500";
    case "medium":
      return "text-yellow-500";
    case "low":
      return "text-gray-500";
  }
}

export function getPriorityBgColor(priority: RecommendationPriority): string {
  switch (priority) {
    case "critical":
      return "bg-red-100 text-red-800";
    case "high":
      return "bg-orange-100 text-orange-800";
    case "medium":
      return "bg-yellow-100 text-yellow-800";
    case "low":
      return "bg-gray-100 text-gray-800";
  }
}

export function getTypeIcon(type: RecommendationType): string {
  switch (type) {
    case "contact_action":
      return "user";
    case "deal_action":
      return "briefcase";
    case "follow_up":
      return "clock";
    case "outreach":
      return "send";
    case "engagement":
      return "message-circle";
    case "upsell":
      return "trending-up";
    case "retention":
      return "shield";
    case "prioritization":
      return "list";
    default:
      return "lightbulb";
  }
}

export function getTypeLabel(type: RecommendationType): string {
  switch (type) {
    case "contact_action":
      return "Contact Action";
    case "deal_action":
      return "Deal Action";
    case "follow_up":
      return "Follow Up";
    case "outreach":
      return "Outreach";
    case "engagement":
      return "Engagement";
    case "upsell":
      return "Upsell";
    case "retention":
      return "Retention";
    case "prioritization":
      return "Priority";
    default:
      return "Recommendation";
  }
}

export function formatScore(score: number): string {
  return `${Math.round(score * 100)}%`;
}

File: backend/app/models/recommendations.py
"""Models for AI-powered recommendations and next-best-actions."""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4

from pydantic import BaseModel as PydanticBaseModel
from sqlalchemy import Column
from sqlalchemy.dialects.postgresql import JSONB
from sqlmodel import Field, SQLModel


class RecommendationType(str, Enum):
    """Types of recommendations."""
    CONTACT_ACTION = "contact_action"  # Actions for specific contacts
    DEAL_ACTION = "deal_action"  # Actions for specific deals
    FOLLOW_UP = "follow_up"  # Follow-up reminders
    OUTREACH = "outreach"  # Outreach suggestions
    ENGAGEMENT = "engagement"  # Engagement opportunities
    UPSELL = "upsell"  # Upsell/cross-sell opportunities
    RETENTION = "retention"  # Retention actions
    PRIORITIZATION = "prioritization"  # Task prioritization


class RecommendationPriority(str, Enum):
    """Priority levels for recommendations."""
    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class RecommendationStatus(str, Enum):
    """Status of a recommendation."""
    PENDING = "pending"
    ACCEPTED = "accepted"
    DISMISSED = "dismissed"
    COMPLETED = "completed"
    EXPIRED = "expired"


class Recommendation(SQLModel, table=True):
    """Database model for AI recommendations."""

    __tablename__ = "recommendations"

    id: UUID = Field(default_factory=uuid4, primary_key=True)

    # Target entity
    contact_id: Optional[UUID] = Field(default=None, foreign_key="contacts.id", index=True)
    deal_id: Optional[UUID] = Field(default=None, foreign_key="deals.id", index=True)

    # Recommendation details
    type: RecommendationType = Field(index=True)
    priority: RecommendationPriority = Field(default=RecommendationPriority.MEDIUM, index=True)
    status: RecommendationStatus = Field(default=RecommendationStatus.PENDING, index=True)

    # Content
    title: str = Field(max_length=200)
    description: str
    reasoning: str  # Why this recommendation was made

    # Action details
    suggested_action: str  # Specific action to take
    action_template: Optional[str] = None  # Template for email/message if applicable

    # Context and scoring
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    impact_score: float = Field(default=0.0, ge=0.0, le=1.0)  # Expected impact
    urgency_score: float = Field(default=0.0, ge=0.0, le=1.0)  # Time sensitivity

    # Supporting data
    context_data: dict = Field(default_factory=dict, sa_column=Column(JSONB))

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    expires_at: Optional[datetime] = None
    acted_on_at: Optional[datetime] = None

    # AI metadata
    model_version: str = Field(default="llama3.1")


# Response models
class RecommendationResponse(PydanticBaseModel):
    """API response model for a recommendation."""
    id: UUID
    contact_id: Optional[UUID]
    deal_id: Optional[UUID]
    type: RecommendationType
    priority: RecommendationPriority
    status: RecommendationStatus
    title: str
    description: str
    reasoning: str
    suggested_action: str
    action_template: Optional[str]
    confidence: float
    impact_score: float
    urgency_score: float
    context_data: dict
    created_at: datetime
    expires_at: Optional[datetime]

    class Config:
        from_attributes = True


class RecommendationListResponse(PydanticBaseModel):
    """API response for list of recommendations."""
    recommendations: list[RecommendationResponse]
    total: int
    pending_count: int
    high_priority_count: int


class RecommendationActionRequest(PydanticBaseModel):
    """Request to act on a recommendation."""
    action: str  # "accept", "dismiss", "complete"
    notes: Optional[str] = None


class NextBestActionsResponse(PydanticBaseModel):
    """Response with prioritized next-best-actions."""
    contact_actions: list[RecommendationResponse]
    deal_actions: list[RecommendationResponse]
    follow_ups: list[RecommendationResponse]
    top_priorities: list[RecommendationResponse]
    generated_at: datetime
